File: SDCloudUserDefaults/SDCloudUserDefaults.m
//
//  SDCloudUserDefaults.m
//
//  Created by Stephen Darlington on 01/09/2011.
//  Copyright (c) 2011 Wandle Software Limited. All rights reserved.
//

#import "SDCloudUserDefaults.h"

@implementation SDCloudUserDefaults

NSString * const SDCloudValueUpdatedNotification = @"com.wandlesoftware.SDCloudUserDefaults.KeyValueUpdated";

static id notificationObserver;
static NSString *suiteName;
static NSUserDefaults *userDefaults;

+ (void) setiCloudEnabled:(BOOL)iCloudEnabled {
    [[NSUserDefaults standardUserDefaults] setBool:iCloudEnabled forKey:ICLOUD_DATA_ENABLED_KEY];
    [[NSUserDefaults standardUserDefaults] synchronize];
    
    if (iCloudEnabled == YES) {
        [SDCloudUserDefaults registerForNotifications];
    } else {
        [SDCloudUserDefaults removeNotifications];
    }
}

+ (BOOL) isiCloudEnabled {
    return [[NSUserDefaults standardUserDefaults] boolForKey:ICLOUD_DATA_ENABLED_KEY];
}

+ (void) initialize {
    [super initialize];
    
    id keyValue = [SDCloudUserDefaults objectForKey:ICLOUD_DATA_ENABLED_KEY];
    
    if (keyValue != nil) {
        BOOL iCloudEnabled = [SDCloudUserDefaults boolForKey:ICLOUD_DATA_ENABLED_KEY];
        
        if (iCloudEnabled == YES) {
            [SDCloudUserDefaults registerForNotifications];
        } else {
            [SDCloudUserDefaults removeNotifications];
        }
    }
}

+(NSUserDefaults *) _standardUserDefaults {
    if (userDefaults == nil) {
        userDefaults = [NSUserDefaults standardUserDefaults];
    }
    return userDefaults;
}

+(void)setSuiteName:(NSString *)aSuiteName {
    if (suiteName == aSuiteName) {
        return;
    }
#if __has_feature(objc_arc)
    suiteName = aSuiteName;
#else
    [suiteName release];
    suiteName = [aSuiteName copy];
#endif
    if (!suiteName) {
#if !__has_feature(objc_arc)
        [userDefaults release];
#endif
        userDefaults = nil;
    }
    else {
        userDefaults = [[NSUserDefaults alloc] initWithSuiteName:suiteName];
        NSDictionary *dictionary = [[NSUbiquitousKeyValueStore defaultStore] dictionaryRepresentation];
        for (NSString *key in [dictionary allKeys]) {
            if (![userDefaults objectForKey:key]) {
                [userDefaults setObject:dictionary[key] forKey:key];
            }
        }
    }
}

+(NSString*)suiteName {
    return suiteName;
}

+(NSString*)stringForKey:(NSString*)aKey {
    return [SDCloudUserDefaults objectForKey:aKey];
}

+(BOOL)boolForKey:(NSString*)aKey {
    return [[SDCloudUserDefaults objectForKey:aKey] boolValue];
}

+(id)objectForKey:(NSString*)aKey {
    if ([SDCloudUserDefaults isiCloudEnabled] == YES) {
        NSUbiquitousKeyValueStore* cloud = [NSUbiquitousKeyValueStore defaultStore];
        id retv = [cloud objectForKey:aKey];
       
        if (!retv) {
            retv = [[self _standardUserDefaults] objectForKey:aKey];
            [cloud setObject:retv forKey:aKey];
        }
        
        return retv;
    } else {
        return [[self _standardUserDefaults] objectForKey:aKey];
    }
}

+(NSInteger)integerForKey:(NSString*)aKey {
    return [[SDCloudUserDefaults objectForKey:aKey] integerValue];
}

+(float)floatForKey:(NSString *)aKey {
    return [[SDCloudUserDefaults objectForKey:aKey] floatValue];
}

+(void)setString:(NSString*)aString forKey:(NSString*)aKey {
    [SDCloudUserDefaults setObject:aString forKey:aKey];
}

+(void)setBool:(BOOL)aBool forKey:(NSString*)aKey {
    [SDCloudUserDefaults setObject:[NSNumber numberWithBool:aBool] forKey:aKey];
}

+(void)setObject:(id)anObject forKey:(NSString*)aKey {
    if ([SDCloudUserDefaults isiCloudEnabled] == YES) {
        [[NSUbiquitousKeyValueStore defaultStore] setObject:anObject forKey:aKey];
    }
    
    [[self _standardUserDefaults] setObject:anObject forKey:aKey];
    
    [SDCloudUserDefaults synchronize];
}

+(void)setInteger:(NSInteger)anInteger forKey:(NSString*)aKey {
    [SDCloudUserDefaults setObject:[NSNumber numberWithInteger:anInteger]
                            forKey:aKey];
}

+(void)setFloat:(float)aFloat forKey:(NSString *)aKey {
    [SDCloudUserDefaults setObject:@(aFloat)
                            forKey:aKey];
}

+(void)removeObjectForKey:(NSString*)aKey {
    if ([SDCloudUserDefaults isiCloudEnabled] == YES) {
        [[NSUbiquitousKeyValueStore defaultStore] removeObjectForKey:aKey];
    }
    
    [[self _standardUserDefaults] removeObjectForKey:aKey];
    
    [SDCloudUserDefaults synchronize];
}

+(void)synchronize {
    if ([SDCloudUserDefaults isiCloudEnabled] == YES) {
        [[NSUbiquitousKeyValueStore defaultStore] synchronize];
    }
    
    [[self _standardUserDefaults] synchronize];
}

+(void)registerForNotifications {
    @synchronized(notificationObserver) {
        if (notificationObserver) {
            return;
        }

        notificationObserver = [[NSNotificationCenter defaultCenter]
                                addObserverForName:NSUbiquitousKeyValueStoreDidChangeExternallyNotification
                                object:[NSUbiquitousKeyValueStore defaultStore]
                                queue:nil
                                usingBlock:^(NSNotification* notification) {
                                    
                                    NSDictionary* userInfo = [notification userInfo];
                                    NSNumber* reasonForChange = [userInfo objectForKey:NSUbiquitousKeyValueStoreChangeReasonKey];
                                    
                                    // If a reason could not be determined, do not update anything.
                                    if (!reasonForChange)
                                        return;
                                    
                                    // Update only for changes from the server.
                                    NSInteger reason = [reasonForChange integerValue];
                                    if ((reason == NSUbiquitousKeyValueStoreServerChange) ||
                                        (reason == NSUbiquitousKeyValueStoreInitialSyncChange)) {
                                        
                                        NSUserDefaults* defaults = [self _standardUserDefaults];
                                        NSUbiquitousKeyValueStore* cloud = [NSUbiquitousKeyValueStore defaultStore];
                                        NSArray* changedKeys = [userInfo objectForKey:NSUbiquitousKeyValueStoreChangedKeysKey];
                                        
                                        for (NSString* key in changedKeys) {
                                            id obj = [cloud objectForKey:key];

                                            [defaults setObject:obj forKey:key];

                                            if (obj != nil) {
                                                [[NSNotificationCenter defaultCenter] postNotificationName:SDCloudValueUpdatedNotification
                                                                                                    object:self
                                                                                                  userInfo:@{key:[cloud objectForKey:key]}];
                                            } else {
                                                [[NSNotificationCenter defaultCenter] postNotificationName:SDCloudValueUpdatedNotification
                                                                                                    object:self
                                                                                                  userInfo:nil];
                                            }
                                        }
                                        
                                        
                                    }
                                }];
    }
    
}

+(void)removeNotifications {
    @synchronized(notificationObserver) {
        [[NSNotificationCenter defaultCenter] removeObserver:notificationObserver];
        notificationObserver = nil;
    }
}

@end

File: SDCloudUserDefaults_Swift/SDCloudUserDefaults.swift
//
//  SDCloudUserDefaults.swift
//  Tap Times Tables Swift
//
//  Created by Peter Easdown on 9/10/20.
//

import Foundation

extension Notification.Name {

    static let SDCloudValueUpdatedNotification = Notification.Name(rawValue: "com.wandlesoftware.SDCloudUserDefaults.KeyValueUpdated")

}

class SDCloudUserDefaults {

    static let ICLOUD_DATA_ENABLED_KEY = "iCloudDataEnabled"
    
    static let shared = SDCloudUserDefaults()

    private var userDefaults : UserDefaults? = nil

    func setiCloud(enabled: Bool) {
        UserDefaults.standard.set(enabled, forKey: SDCloudUserDefaults.ICLOUD_DATA_ENABLED_KEY)
    
        if enabled {
            self.registerForNotifications()
        } else {
            self.removeNotifications()
        }
    }

    func hasiCloudBeenInitialized() -> Bool {
        return UserDefaults.standard.object(forKey: SDCloudUserDefaults.ICLOUD_DATA_ENABLED_KEY) != nil
    }

    func isiCloudEnabled() -> Bool {
        return UserDefaults.standard.bool(forKey: SDCloudUserDefaults.ICLOUD_DATA_ENABLED_KEY)
    }

    init() {
        if self.hasiCloudBeenInitialized() {
            if self.isiCloudEnabled() {
                self.registerForNotifications()
            } else {
                self.removeNotifications()
            }
        }
    }
    
    func standardUserDefaults() -> UserDefaults {
        if self.userDefaults == nil {
            self.userDefaults = UserDefaults.standard
        }
        
        return self.userDefaults!
    }

    func string(forKey aKey: String) -> String? {
        if let obj = self.object(forKey: aKey) {
            if obj is String {
                return obj as? String
            }
        }
        
        return nil
    }
    
    func bool(forKey aKey: String) -> Bool {
        if let obj = self.object(forKey: aKey) {
            if obj is Bool {
                return obj as! Bool
            }
        }
        
        return false
    }
    
    func object(forKey aKey: String) -> Any? {
        if self.isiCloudEnabled() {
            if let obj = NSUbiquitousKeyValueStore.default.object(forKey: aKey) {
                return obj
            } else {
                if let obj = self.standardUserDefaults().object(forKey: aKey) {
                    NSUbiquitousKeyValueStore.default.set(obj, forKey: aKey)
                    
                    return obj
                } else {
                    return nil
                }
            }
        } else {
            return self.standardUserDefaults().object(forKey: aKey)
        }
    }

    func array(forKey aKey: String) -> Array<Any>? {
        if let obj = self.object(forKey: aKey) {
            if obj is Array<Any> {
                return obj as? Array<Any>
            }
        }
        
        return nil
    }

    func data(forKey aKey: String) -> Data? {
        if let obj = self.object(forKey: aKey) {
            if obj is Data {
                return obj as? Data
            }
        }
        
        return nil
    }

    func integer(forKey aKey: String) -> Int {
        if let obj = self.object(forKey: aKey) {
            if obj is Int {
                return obj as! Int
            }
        }
        
        return 0
    }

    func float(forKey aKey: String) -> Float {
        if let obj = self.object(forKey: aKey) {
            if obj is Float {
                return obj as! Float
            }
        }
        
        return 0.0
    }

    func double(forKey aKey: String) -> Double {
        if let obj = self.object(forKey: aKey) {
            if obj is Double {
                return obj as! Double
            }
        }
        
        return 0.0
    }

    func set(string aString: String, forKey aKey: String) {
        self.set(aString, forKey:aKey)
    }

    func set(_ aBool: Bool, forKey aKey: String) {
        self.set(NSNumber(booleanLiteral: aBool), forKey: aKey)
    }

    func set(_ object: Any, forKey aKey: String) {
        if self.isiCloudEnabled() {
            NSUbiquitousKeyValueStore.default.set(object, forKey: aKey)
        }

        self.userDefaults?.set(object, forKey: aKey)
    }

    func set(_ integer: Int, forKey aKey: String) {
        self.set(NSNumber(integerLiteral: integer), forKey: aKey)
    }

    func set(_ float: Float, forKey aKey: String) {
        self.set(NSNumber(floatLiteral: Double(float)), forKey: aKey)
    }

    func set(_ double: Double, forKey aKey: String) {
        self.set(NSNumber(floatLiteral: double), forKey: aKey)
    }

    func removeObject(forKey aKey: String) {
        if self.isiCloudEnabled() {
            NSUbiquitousKeyValueStore.default.removeObject(forKey: aKey)
        }
        
        self.userDefaults?.removeObject(forKey: aKey)
    }

    func synchronize() {
        if self.isiCloudEnabled() {
            NSUbiquitousKeyValueStore.default.synchronize()
        }
        
        self.userDefaults?.synchronize()
    }
    
    var notificationObserver : NSObjectProtocol? = nil

    func registerForNotifications() {
        if self.notificationObserver != nil {
            return
        }
        
        self.notificationObserver = NotificationCenter.default.addObserver(forName: NSUbiquitousKeyValueStore.didChangeExternallyNotification, object: NSUbiquitousKeyValueStore.default, queue: nil, using: { (notification) in
            if let userInfo = notification.userInfo {
                if let reasonForChange = userInfo[NSUbiquitousKeyValueStoreChangeReasonKey] {
                    let reason = (reasonForChange as! NSNumber).intValue
                    
                    // Update only for changes from the server.
                    if ((reason == NSUbiquitousKeyValueStoreServerChange) ||
                            (reason == NSUbiquitousKeyValueStoreInitialSyncChange)) {
                        
                        let changedKeys = userInfo[NSUbiquitousKeyValueStoreChangedKeysKey] as! Array<String>
                        let cloud = NSUbiquitousKeyValueStore.default
                        let defaults = self.userDefaults
                        
                        changedKeys.forEach { (key) in
                            if let obj = cloud.object(forKey: key) {
                                defaults?.setValue(obj, forKey: key)
                                NotificationCenter.default.post(name: .SDCloudValueUpdatedNotification, object: self, userInfo: [key: obj])
                            }
                        }
                    }
                }
            }
        })
    }
    
    func removeNotifications() {
        NotificationCenter.default.removeObserver(self.notificationObserver!)
        self.notificationObserver = nil
    }
}
